---
title: "ASCAT to RaggedExperiment"
author:
- name: Lydia King
  affiliation: University of Galway, Ireland
- name: Marcel Ramos
  affiliation: Roswell Park Cancer Institute, Buffalo, NY
date: "`r BiocStyle::doc_date()`"
vignette: |
  %\VignetteIndexEntry{ASCAT to RaggedExperiment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    toc_float: true
Package: RaggedExperiment
---

# Introduction

The [*RaggedExperiment*](https://bioconductor.org/packages/3.16/RaggedExperiment) package provides a flexible data representation for copy number, mutation and other ragged array schema for genomic location data. The output of Allele-Specific Copy number Analysis of Tumors (ASCAT) can be classed as a ragged array and contains whole genome allele-specific copy number information for each sample in the analysis. For further analysis of the ASCAT data, utilizing the functionalities of `RaggedExperiment`, the ASCAT data must undergo a number of operations to get it in the correct format for use with `RaggedExperiment`. 

# Installation

```{r, message = FALSE, warning = FALSE}
if (!require("BiocManager"))
    install.packages("BiocManager")
BiocManager::install("RaggedExperiment")
```

Loading the package:

```{r, message = FALSE}
library(RaggedExperiment)
library(GenomicRanges)
```

# Structure of ASCAT data

The data shown below is the output obtained from ASCAT. ASCAT takes Log R Ratio (LRR) and B Allele Frequency (BAF) files and derives the allele-specific copy number profiles of tumour cells, accounting for normal cell admixture and tumour aneuploidy. It should be noted that if working with raw CEL files, the first step is to preprocess the CEL files using the PennCNV-Affy pipeline described [here](https://penncnv.openbioinformatics.org/en/latest/user-guide/affy/). The PennCNV-Affy pipeline produces the LRR and BAF files used as inputs for ASCAT.  

Let's load up and have a look at the output obtained from ASCAT. Here we look at the first 10 rows of the dataframe. 

```{r}
ASCAT_data <- read.csv("../data/ASCAT_Sample_File.txt", sep="\t")

head(ASCAT_data, n = 10)
```

From the output above we can see that the ASCAT data has 6 columns named sample, chr, startpos, endpos, nMajor and nMinor. These correspond to the sample ID, chromosome, the start position and end position of the genomic ranges and the copy number of the major and minor alleles i.e. the homologous chromosomes.  

The ASCAT data provides information for all samples processed in an ASCAT run in 1 dataframe. In this case our data contains copy number information for three samples, sample1, sample2 and sample3.

```{r}
unique(ASCAT_data$sample)
```

# Converting ASCAT data to `GRanges` format

The `RaggedExperiment` class derives from a `GRangesList` representation and can take a `GRanges` object, a `GRangesList` or a list of `Granges` as inputs. To be able to use the ASCAT data in `RaggedExperiment` we must convert the ASCAT data into `GRanges` format. Ideally, we want each of our `GRanges` objects to correspond to an individual sample.

## ASCAT to `GRanges` objects

In the case where the ASCAT data has only 1 sample it is relatively simple to produce a `GRanges` object.

```{r}
ASCAT_data_1sample <- ASCAT_data[ASCAT_data$sample == "sample1",]

sample1_ex1 <- GRanges(seqnames = Rle(paste0("chr", ASCAT_data_1sample$chr)),
    ranges = IRanges(start = ASCAT_data_1sample$startpos, end = ASCAT_data_1sample$endpos),
    strand = Rle(strand(c("*"))),
    nmajor = ASCAT_data_1sample$nMajor,
    nminor = ASCAT_data_1sample$nMinor)

sample1_ex1
```

Here we create a `GRanges` object by taking each column of the ASCAT data and assigning them to the appropriate argument in the `GRanges` function. From above we can see that the chromosome information is prefixed with "chr" and becomes the seqnames column, the start and end positions are combined into an `IRanges` object and given to the ranges argument, the strand column contains a `*` for each entry as we don't have strand information and the metadata columns contain the allele-specific copy number calls and are called nmajor and nminor. The `GRanges` object we have just created contains 41 ranges (rows) and 2 metadata columns.

Another way that we can easily convert our ASCAT data, containing 1 sample, to a `GRanges` object is to use the `makeGRangesFromDataFrame` function from the `GenomicsRanges` package. Here we indicate what columns in our data correspond to the chromosome (given to the `seqnames` argument), start and end positions (`start.field` and `end.field` arguments), whether to ignore strand information and assign all entries `*` (`ignore.strand`) and also whether to keep the other columns in the dataframe, nmajor and nminor, as metadata columns (`keep.extra.columns`).

```{r}
sample1_ex2 <- makeGRangesFromDataFrame(ASCAT_data_1sample[,-c(1)], ignore.strand=TRUE, seqnames.field=c("chr"), start.field="startpos", end.field=c("endpos"), keep.extra.columns=TRUE)

sample1_ex2
```

In the case where the ASCAT data contains more than 1 sample we must first split the dataframe based on sample and then create a `GRanges` object for each sample. This can be done by splitting the dataframe and subsequently creating a `GRanges` object for each sample, or by using the `makeGRangesListFromDataFrame` function from the `GenomicsRanges` package. Here we show how to split the whole dataframe into a dataframe for each sample and use the `makeGRangesFromDataFrame` to create our `GRanges` objects. In the next section we will show how to use the `makeGRangesListFromDataFrame` function.

```{r}
sample_list <- split(ASCAT_data, f = ASCAT_data$sample) 

sample1_ex3 <- makeGRangesFromDataFrame(sample_list[[1]][,-c(1)], ignore.strand=TRUE, seqnames.field=c("chr"), start.field="startpos", end.field=c("endpos"), keep.extra.columns=TRUE)

sample1_ex3

sample2_ex3 <- makeGRangesFromDataFrame(sample_list[[2]][,-c(1)], ignore.strand=TRUE, seqnames.field=c("chr"), start.field="startpos", end.field=c("endpos"), keep.extra.columns=TRUE)

sample2_ex3

sample3_ex3 <- makeGRangesFromDataFrame(sample_list[[3]][,-c(1)], ignore.strand=TRUE, seqnames.field=c("chr"), start.field="startpos", end.field=c("endpos"), keep.extra.columns=TRUE)

sample3_ex3
```

## ASCAT to `GRangesList` instance

To produce a `GRangesList` instance from the ASCAT dataframe we can use the `makeGRangesListFromDataFrame` function. This function can take the same arguments as the `makeGRangesFromDataFrame` function used above, but also has an argument specifying how the rows of the `df` are split (split.field). Here we will split on sample. This function can be used in cases where the ASCAT data contains only 1 sample or where it contains multiple samples.

Using `makeGRangesListFromDataFrame` to create a list of `GRanges` objects where ASCAT data only has 1 sample:  

```{r}
sample_list_GRanges_ex1 <- makeGRangesListFromDataFrame(ASCAT_data_1sample, ignore.strand=TRUE, seqnames.field=c("chr"), start.field="startpos", end.field=c("endpos"), keep.extra.columns=TRUE, split.field = "sample")

sample_list_GRanges_ex1
```

Using `makeGRangesListFromDataFrame` to create a list of `GRanges` objects where ASCAT data has multiple samples:

```{r}
sample_list_GRanges_ex2 <- makeGRangesListFromDataFrame(ASCAT_data, ignore.strand=TRUE, seqnames.field=c("chr"), start.field="startpos", end.field=c("endpos"), keep.extra.columns=TRUE, split.field = "sample")

sample_list_GRanges_ex2
```

Each `GRanges` object in the `list` can then be accessed using square bracket notation.

```{r}
sample1_ex4 <- sample_list_GRanges_ex2[[1]]

sample1_ex4
```

## ASCAT to `list` of `GRanges`

Another thing we might want to do is create a list that contains all of our `GRanges` objects. This is straightforward in that we use the `list` function with our `GRanges` objects as named or unnamed inputs. Below we have created a list that includes 3 `GRanges` objects, created at the end of section 4.1., corresponding to our 3 samples.

```{r}
rangeList <- list(sample1 = sample1_ex3, sample2 = sample2_ex3, sample3 = sample3_ex3)

rangeList
```

# Constructing a `RaggedExperiment` object from ASCAT output

Now we have created the `GRanges` object, `GRangesList` and `list` of `GRanges` we can easily use `RaggedExperiment`.

## Using `GRanges` objects

From above we have a `GRanges` object derived from the ASCAT data containing 1 sample i.e. `sample1_ex1` / `sample1_ex2` and the individual `GRanges` objects derived from the ASCAT data containing 3 samples i.e. `sample1_ex3`, `sample2_ex3` and `sample3_ex3`. We can now use these `GRanges` objects as inputs to `RaggedExperiment`. Note that we create column data `colData` to describe the samples.

Using `GRanges` object where ASCAT data only has 1 sample:

```{r}
colDat_1 = DataFrame(id = 1)
ragexp <- RaggedExperiment(sample1 = sample1_ex2, 
                           colData = colDat_1)
ragexp
```

Using `GRanges` object where ASCAT data only has multiple samples:

```{r}
colDat_3 = DataFrame(id = 1:3)
ragexp <- RaggedExperiment(sample1 = sample1_ex3,
                           sample2 = sample2_ex3,
                           sample3 = sample3_ex3, 
                           colData = colDat_3)
ragexp
```

## Using a `GRangesList` instance

From before we have a `GRangesList` derived from the ASCAT data containing 1 sample i.e. `sample_list_GRanges_ex1` and the `GRangesList` derived from the ASCAT data containing 3 samples i.e. `sample_list_GRanges_ex2`. We can now use this `GRangesList` as the input to `RaggedExperiment`.

Using `GRangesList` where ASCAT data only has 1 sample:

```{r}
RaggedExperiment(sample_list_GRanges_ex1, colData = colDat_1)
```

Using `GRangesList` where ASCAT data only has multiple samples:

```{r}
RaggedExperiment(sample_list_GRanges_ex2, colData = colDat_3)
```

## Using a `list` of `GRanges`

From above we have a `list` of `GRanges` i.e. `rangeList` containing 3 `GRanges` objects. We can now use this `list` of `GRanges` as the input to `RaggedExperiment`.

```{r}
RaggedExperiment(rangeList, colData = colDat_3)
```

## Using a `List` of `GRanges` with metadata

There may be cases where a `SimpleGenomicRangesList` is provided that has corresponding metadata (accessed by `mcols`). In these cases the metadata is used as the `colData` for the `RaggedExperiment`. Here we use the `List` function in the same way as we would use the `list` function (shown above), but with the metadata being added using the `mcols` function.

```{r}
grList <- List(sample1 = sample1_ex3, sample2 = sample2_ex3, sample3 = sample3_ex3)
mcols(grList) <- colDat_3

RaggedExperiment(grList)
```

# Common use cases (Applications?)
